pycloudlib/oci/cloud.py

```diff
@@ -0,0 +1,227 @@
+# This file is part of pycloudlib. See LICENSE file for license information.
+"""OCI Cloud type."""
+
+import os
+import re
+import oci
+
+from pycloudlib.cloud import BaseCloud
+from pycloudlib.oci.instance import OciInstance
+from pycloudlib.oci.utils import wait_till_ready
+from pycloudlib.util import UBUNTU_RELEASE_VERSION_MAP
+
+
+class OCI(BaseCloud):
+    """OCI (Oracle) cloud class."""
+
+    _type = 'oci'
+
+    def __init__(self, tag, compartment_id, config_path='~/.oci/config'):
+        """
+        Initialize the connection to OCI.
+
+        OCI must be initialized on the CLI first:
+        https://github.com/cloud-init/qa-scripts/blob/master/doc/launching-oracle.md
+
+        Args:
+            tag: Name of instance
+            compartment_id: A compartment found at
+                https://console.us-phoenix-1.oraclecloud.com/a/identity/compartments
+            config_path: Path of OCI config file
+        """
+        super(OCI, self).__init__(tag)
+        self.compartment_id = compartment_id
+
+        if not os.path.isfile(os.path.expanduser(config_path)):
+            raise ValueError(
+                '{} is not a valid config file. '
+                'Pass a valid config file or first setup your OCI client. '
+                'See https://github.com/cloud-init/qa-scripts/blob/master/'
+                'doc/launching-oracle.md'
+                .format(config_path))
+        self.config_path = config_path
+        config = oci.config.from_file(str(config_path))
+
+        self._log.debug('Logging into OCI')
+        self.compute_client = oci.core.ComputeClient(config)
+        self.network_client = oci.core.VirtualNetworkClient(config)
+
+    def delete_image(self, image_id):
+        """Delete an image.
+
+        Args:
+            image_id: string, id of the image to delete
+        """
+        self.compute_client.delete_image(image_id)
+
+    def released_image(self, release, operating_system='Canonical Ubuntu'):
+        """Get the released image.
+
+        OCI just has periodic builds, so "released" and "daily" don't
+        really make sense here. Just call the same code for both
+
+        Args:
+            release: string, Ubuntu release to look for
+            operating_system: string, operating system to use
+        Returns:
+            string, id of latest image
+
+        """
+        return self.daily_image(release, operating_system)
+
+    def daily_image(self, release, operating_system='Canonical Ubuntu'):
+        """Get the daily image.
+
+        OCI just has periodic builds, so "released" and "daily" don't
+        really make sense here. Just call the same code for both
+
+        Args:
+            release: string, Ubuntu release to look for
+            operating_system: string, Operating system to use
+
+        Returns:
+            string, id of latest image
+
+        """
+        if operating_system == 'Canonical Ubuntu':
```

Comment:
Not a blocker to release of this PR, What in your mind differentiates 'Canonical Ubuntu' from just 'ubuntu'? I'm thinking about things in terms of when your followup refactor-api tries to generalize methods up in pycloudlib.cloud parent class, and ease of use when folks try to run 'ubuntu' vs. 'opensuse' 'debian' or others. 

I realize this might just be an Oracle CLI need, for which I'd say we'd probably like to standardize on something simple in the parent class, that get's translated to "Canonical Ubuntu" in the oracle api call

Reply:
This is just a consequence of the Oracle API. If I want to list Ubuntu images with the `self.compute_client.list_images(...)` call, I have to pass 'Canonical Ubuntu' as the `operating_system` argument. In my head, this was an Oracle-specific thing. In my refactor-api branch, I was expecting the method to look like `def daily_image(self, release, **kwargs)`. Since the default is defined here, that's how the API will normally function, but if you have a reason to make some Oracle-specific calls, you can override which operating_system selection.